modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/RuntimeHashIndex.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ignite.internal.processors.query.calcite.exec;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Supplier;
+
+import org.apache.calcite.util.ImmutableBitSet;
+import org.apache.ignite.internal.processors.query.calcite.exec.exp.agg.GroupKey;
+import org.apache.ignite.internal.util.typedef.F;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Runtime sorted index based on on-heap tree.
+ */
+public class RuntimeHashIndex<Row> implements RuntimeIndex<Row> {
+    /** */
+    protected final ExecutionContext<Row> ectx;
+
+    /** */
+    private final ImmutableBitSet keys;
+
+    /** Rows. */
+    private HashMap<GroupKey, List<Row>> rows;
+
+    /**
+     *
+     */
+    public RuntimeHashIndex(
+        ExecutionContext<Row> ectx,
+        ImmutableBitSet keys
+    ) {
+        this.ectx = ectx;
+
+        assert !F.isEmpty(keys);
+
+        this.keys = keys;
+        rows = new HashMap<>();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void push(Row r) {
+        List<Row> newEqRows = new ArrayList<>();
```

Comment:
why do u need this gc pressure here ? 
List<Row> eqRows = rows.putIfAbsent(key(r), new ArrayList<>()); 
wdyt?